7_zebra/zebra_pulp.py

```diff
@@ -0,0 +1,126 @@
+"""
+Solution to the Zebra puzzle.
+
+This version uses PuLP as a modeling language and CBC as a solver.
+
+Created by Éder Pinheiro and Eric Zettermann  (Sep 9, 2021), MipMaster.org.
+"""
+
+import pulp
+
+# region Input Data
+# house positions
+I = [1, 2, 3, 4, 5]
+# attributes
+J = ['flag', 'color', 'pet', 'beverage', 'hobby']
+# options
+K = {'flag': ['Brazil', 'Venezuela', 'India', 'Tunisia', 'Mexico'],
+     'color': ['red', 'green', 'ivory', 'yellow', 'blue'],
+     'pet': ['dog', 'cat', 'turtle', 'hamster', 'rabbit'],
+     'beverage': ['coffee', 'tea', 'milk', 'orange juice', 'water'],
+     'hobby': ['chess', 'volleyball', 'sudoku', 'karaoke', 'fishing']}
+
+# keys for decision variables x
+x_keys = [(i, j, k) for i in I for j in J for k in K[j]]
+# endregion
+
+# region Define the model
+mdl = pulp.LpProblem('zebra', sense=pulp.LpMaximize)
+
+# add variables
+x = pulp.LpVariable.dicts(indexs=x_keys, cat=pulp.LpBinary, name='x')
+
+# add constraints
+# exactly one option for each house-attribute pair
+for i in I:
+    for j in J:
+        mdl.addConstraint(pulp.lpSum(x[i, j, k] for k in K[j]) == 1, name=f'single_option_{i}_{j}')
+
+# exactly one house for each attribute-option pair
+for j in J:
+    for k in K[j]:
+        mdl.addConstraint(pulp.lpSum(x[i, j, k] for i in I) == 1, name=f'single_house_{j}_{k}')
+
+# R2
+mdl.addConstraint(pulp.lpSum(i * x[i, 'flag', 'Brazil'] for i in I) == pulp.lpSum(i * x[i, 'color', 'red']
+                                                                                     for i in I), name='R2')
+
+# R3
+mdl.addConstraint(pulp.lpSum(i * x[i, 'flag', 'Venezuela'] for i in I) == pulp.lpSum(i * x[i, 'pet', 'dog']
+                                                                                        for i in I), name='R3')
+
+# R4
+mdl.addConstraint(pulp.lpSum(i * x[i, 'color', 'green'] for i in I) == pulp.lpSum(i * x[i, 'beverage', 'coffee']
+                                                                                  for i in I), name='R4')
+# R5
+mdl.addConstraint(pulp.lpSum(i * x[i, 'flag', 'India'] for i in I) == pulp.lpSum(i * x[i, 'beverage', 'tea']
+                                                                                    for i in I), name='R5')
+
+# R6
+for i in I[1:]:
+    mdl.addConstraint(x[i, 'color', 'green'] == x[i - 1, 'color', 'ivory'], name=f'R6_{i}')
+# mdl.addConstraint(x[1, 'color', 'green'] == 0, name=f'R6_{1}')
```

Comment:
@ericztt , I think that if we use equality here instead of inequality, we don't need the extra constraint in line 62.
Did you have any particular reason to use inequality instead of equality in this constraint?
The same question also applies to constraints R11 and R12.

Reply:
I guess you are right. I used inequalities because I was thinking in if - then implementation.  But as the reciprocate is also true the equality fits better. 

For R11 and R12  we could also use equalities. But we might have to use and extra restriction or a x.get, I am afraid.